Spring Web으로 구현한 ToDo REST API 테스트 코드 작성하기

injae7034 opened the pull request:
1. HelloController 유닛 테스트 -> 완료
2. HelloController MockMVC 테스트 -> 완료
3. Task 모델 테스트 -> 완료
4. TaskService 테스트 -> 완료
5. TaskController 유닛 테스트 -> 완료
6. TaskController MockMVC 테스트 -> 미완성(create(POST)에서 방법을 몰라 멈춤)

injae7034 committed:
테스트 코드 작성

injae7034 committed:
마지막 띄어쓰기 수정

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/HelloController.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class HelloController {
```
이 클래스에 JavaDoc 주석을 작성해서 기능을 설명해 보세요.

injae7034 replied:
넵~

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/HelloController.java:
```diff
@@ -0,0 +1,12 @@
+    public String sayHello() {
+        return "Hello, world!";
+    }
+}
```
파일 마지막에 개행 문자가 없어서 경고가 나오고 있어요.
![image](https://user-images.githubusercontent.com/1855714/161539635-5042cdf4-5c5a-4949-a097-2d7f8a84b4b3.png)

injae7034 replied:
네 개행하도록 하겠습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/AppTest.java:
```diff
@@ -0,0 +1,15 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class AppTest {
+    @Test
+    void appHasAGreeting() {
+        App classUnderTest = new App();
+        assertNotNull(classUnderTest.getGreeting(), "app should have a greeting");
+        assertEquals(classUnderTest.getGreeting(), "Hello, world!");
+    }
```
이 프로젝트를 처음 보는 사람은 이 테스트만 봐서는 용도를 잘 알 수 없을 것입니다. `@DisplayName`을 사용해서 테스트의 목적을 설명해 보세요.

injae7034 replied:
넵~

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,101 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+class TaskServiceTest {
```
Junit5를 사용해 계층 구조의 테스트 코드를 작성하는 방법을 조사해서 적용해 보세요.

injae7034 replied:
넵 적용하도록 하겠습니다~

johngrib reviewed (commented):
이번주도 힘내서 열심히 해 봅시다!

injae7034 committed:
코드 리뷰 받은 내용 개선

HelloController에 JavaDoc 주석 내용 추가<br>
파일 마지막 개행 추가<br>
AppTest에 DisplayName 어노테이션 추가<br>
TaskServiceTest에 Junit5를 사용해 계층 구조의 테스트 코드를 작성<br>

injae7034 commented:
# TaskController MockMVC 테스트 
어제 create(POST) 테스트 코드를 작성하다가 잘몰라서 멈췄었는데,<br>
오늘 서치했지만 아직 정확한 방법을 찾지 못해 보류함.

# 어제 코드 리뷰 받은 내용을 개선함
## 1. HelloController에 JavaDoc 주석 내용 추가
## 2. 파일 마지막 개행 추가
Intellij에서 파일 마지막에 개행이 되도록 설정함.
## 3. AppTest에 DisplayName 어노테이션 추가하여 설명
## 4. Junit5를 사용해 계층 구조의 테스트 코드를 작성
TaskServiceTest와 TaskControllerTest에 적용함.

### 이번 한 주도 화이팅입니다^^

injae7034 committed:
이중박싱제거

반환값의 자료형이 일치하는데 또 Long으로 쓸데없이 박싱하고 있어서 이를 제거함.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/HelloController.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * 콘솔창에서 클라이언트가 서버에 메세지를 보낼 때<br>
+ * 아무 메세지도 보내지 않고, 주소(localhost:8080)만 보내는 경우에<br>
+ * 클라이언트에 대한 서버의 응답을 처리해주는 클래스입니다.
+ */
+
```
주석이 너무 자세합니다. 이건 과제를 하고 있는 한 사람의 관점에서 작성된 주석인데요, 이 코드의 설계자 관점에서 주석을 작성하는 것이 좋겠어요. 이 코드의 진짜 목적이 무엇인지 곰곰히 생각해 보시고 댓글로 설명해 보세요. 그 다음에 주석을 수정하도록 합시다.

injae7034 replied:
이 코드는 클라이언트와 서버가 제대로 연결되었는지 확인하기 위한 테스트 컨트롤러의 역할을 하고 있습니다.<br>
간단하고 빠르게 테스트하기 위해서 Hello, world!가 화면에 잘출력되는지, 잘출력된다면 서버와 클라이언트가 서로 제대로 연결이 되었다는 것을 의미합니다.<br>
제대로 출력이 되지 않는다면 서버 또는 연결에 문제가 있다는 것을 의미합니다.<br>
따라서 HelloController는 단순히 이를 확인하기 위한 Contorller라고 생각합니다.

johngrib replied:
맞습니다! 너무 잘 설명하셨어요. 용도와 의도를 잘 이해하고 계십니다. 그렇다면 이제 정말 필요한 설명만 JavaDoc으로 작성해 봅시다.

johngrib commented on app/src/test/java/com/codesoom/assignment/AppTest.java:
```diff
@@ -0,0 +1,18 @@
+
+@DisplayName("App 클래스")
+class AppTest {
+    @Test
+    @DisplayName("프로그램을 동작시킬 main메소드를 가지고 있는 App클래스가 제대로 생성되었는지 확인합니다.")
+    void appHasAGreeting() {
+        App classUnderTest = new App();
+        assertNotNull(classUnderTest.getGreeting(), "app should have a greeting");
+        assertEquals(classUnderTest.getGreeting(), "Hello, world!");
+    }
```
App 클래스가 생성되었는지는 이 테스트의 목적이 아닙니다. 테스트하려는 것이 `getGreeting`인지 아니면 다른 것인지를 분명히 해야 합니다. 이 테스트를 왜 작성했나요? 생각나는대로 모든 이유를 댓글로 작성해 보세요. 그 중에서 우리가 같이 이야기할만한 거리가 있을 겁니다.

injae7034 replied:
App과 AppTest의 경우 gradle 프로젝트를 생성할 때 디폴트로 있는 것인데<br>
App에서 main메소드는 프로그램을 실행시키기 위해 반드시 필요하지만<br>
getGreeting메소드는 왜 있는지 모르지만 프로젝트를 설정할 때 생성되어 있길래 그냥 뒀었는데<br>
지금 와서 드는 생각은 필요가 없는 것 같습니다.<br>
그래서 getGreeting 메소드는 없애는 것이 맞는 거 같습니다.<br>
그리고 그렇게 되면 AppTest 역시 존재의 이유가 없기 때문에 AppTest도 없애야 한다고 생각합니다.<br>

johngrib replied:
생각대로 해보시고, 결과를 관찰해 보세요. 그런데 생각대로 잘 안 될 수도 있겠죠? 그럴 때를 대비해서 git으로 복구하는 방법을 미리 조사해 두시면 좋을 겁니다.

injae7034 replied:
git reset이나 git revert를 사용하면 될까요?

johngrib replied:
둘 다 괜찮습니다. 히스토리를 남기고 싶다면 revert가 적절하겠죠.

injae7034 replied:
getGreeting 메소드와 AppTest를 지우고 테스트를 계속 하고 있는데 아직 별다른 문제점을 찾지 못하였는데 혹시 나중에 무슨 문제가 생길 수 있나요?

johngrib replied:
테스트 잘 돌아가네요. 문제 없어 보입니다.

![image](https://user-images.githubusercontent.com/1855714/162434668-0b7ddaa2-f6dd-4d4f-8487-2be9cbbe4f7f.png)

injae7034 replied:
전체 테스트 목록까지... 감사합니다^^

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,138 @@
+    @Nested
+    @DisplayName("getTask 메소드는")
+    class Describe_getTask {
+        @Nested
+        @DisplayName("매개변수로 입력 받는 id가 있으면")
+        class Context_with_valid_id {
+            @Test
+            @DisplayName("id에 해당하는 Task를 반환한다.")
+            void getTask() {
+                assertThat(taskService.getTask(1L).getTitle()).isEqualTo(TASK_TITLE_ONE);
+            }
+        }
```
"매개변수로 입력 받는 id" 이야기를 `Context`에서 하고 있죠? 그래서 그 아이디를 `Context`에서 정의해 보도록 합시다. 각 스코프별로 의미를 정의하는 공간이라 생각해 보세요. `Context`는 실험을 준비하는 곳이고, `@Test`는 실험을 실행하는 곳입니다.

injae7034 replied:
id를 정의해 본다는 말이 좀 더 구체적으로 작성을 해야한다는 말씀일까요?<br>
그래서 일단은 "클라이언트가 요청한 Task의 id가 Tasks에 있으면"이라고 수정을 했습니다.

johngrib replied:
`class Context_with_valid_id` 바로 아랫줄에 id를 `final`을 써서 할당해보시라는 뜻이에요. `Context` 스코프의 역할에 대해서 다시 한 번 생각해 봅시다.

injae7034 replied:
아 그렇군요 넵 좀 더 고민해보겠습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMock_MVC_Test.java:
```diff
@@ -0,0 +1,111 @@
+        task.setTitle(TASK_TITLE_ONE);
+        tasks.add(task);
+
+        given(taskService.getTasks()).willReturn(tasks);
+
+        given(taskService.getTask(1L)).willReturn(task);
+
+        given(taskService.getTask(100L)).willThrow(new TaskNotFoundException(100L));
+
```
여기에서는 `given`을 사용할 이유가 없어 보입니다. 아마 테스트 코드를 위해 코딩을 좀 해야 할 거에요. 좀 고생스럽더라도 `given`을 제거해 보세요.

injae7034 replied:
given을 제거하고 직접 taskService에 task를 채운 다음에 테스트를 하라는 말씀이실까요?

johngrib replied:
네. 그렇게도 할 수 있겠죠. 방법은 다양합니다. 처음부터 `given`의 사용에 몰두하지 말고, 일단 테스트를 단단하게 잘 작성하는 걸 연습해 봅시다. given은 그 이후에 연습해도 됩니다. 지금은 given을 쓰는 게 테스트 코드 작성 학습에 안 좋은 영향을 줄 거에요.

injae7034 replied:
given을 사용하지 않는다는 의미가 mockMvc와 그 메소드인 perform도 사용하지 않고, 테스트를 하는 것이 맞을까요?<br>
그렇다면 TaskControllerTest와 TaskControllerMock_MVC_Test의 차이가 무엇인지 여쭤봐도 될까요?<br>
제가 현재 생각나는 차이는  TaskControllerTest는 TaskControllerTest의 메소드 기능이 제대로 돌아가는지를 테스트하는 것으로 생각합니다.<br>
그리고 TaskControllerMock_MVC_Test는 서버를 가동하지 않고도 서버를 가동한 것처럼 TaskController를 테스트하는 것으로 생각하는데, 이 때 필요한 것이 mockMvc이고, 이것을 사용해서 테스트를 하면 서버를 실제로 작동하지 않고도 테스트를 할 수 있는 것이 맞을까요?

johngrib replied:
아뇨. 제가 쓴 코멘트를 다시 읽어보세요. mockMvc 랑 perform 이야기는 한 적이 없어요. given에 대해서만 이야기했죠.
mockMvc는 이번 테스트에서 필수입니다.

그냥 `taskService`에 있는 task 저장 메소드를 호출해서 평범하게 task를 저장한 다음에 호출해 보시라는 뜻입니다.

injae7034 committed:
TaskControllerMock_MVC_Test 개선

injae7034 commented:
## 1. HelloController의 진짜 목적에 대한 고찰 -> O
## 2. AppTest의 존재 이유에 대해 고민  -> O
## 3. 계층 구조의 테스트 구조에서 Context의 내용 좀 더 구체화하기 -> O
## 4. 다른 분들의 코드를 참고하여 TaskControllerMock_MVC_Test에서 POST와 READ, DELETE 코드 변경 -> O
UPDATE는 아직 테스트를 통과하지 못함 -> X

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMock_MVC_Test.java:
```diff
@@ -0,0 +1,173 @@
+                }
+            }
+
+            @Nested
+            @DisplayName("클라이언트가 요청한 Task의 id가 Tasks에 없으면")
+            class Context_with_invalid_id {
+                @Test
+                @DisplayName("TaskNotFoundException 예외를 던집니다.")
+                void detailWithInvalidId() throws Exception {
+                    given(taskService.getTask(100L)).willThrow(new TaskNotFoundException(100L));
+
+                    mockMvc.perform(get("/tasks/100"))
+                            .andExpect(status().isNotFound());
+                }
+            }
```
Context에서 말하고 있는 "상황"을 Context에서 직접 준비하도록 작업해 보세요.
```suggestion
            @Nested
            @DisplayName("클라이언트가 요청한 Task의 id가 Tasks에 없으면")
            class Context_with_invalid_id {

                final long requestTaskId = 100L;  // <--- 이런 식으로 @Test에서 사용할 값을 준비합니다

                @Test
                @DisplayName("TaskNotFoundException 예외를 던집니다.")
                void detailWithInvalidId() throws Exception {
                    given(taskService.getTask(requestTaskId)).willThrow(new TaskNotFoundException(requestTaskId));

                    mockMvc.perform(get("/tasks/" + requestTaskId))
                            .andExpect(status().isNotFound());
                }
            }
```

injae7034 replied:
자세한 설명 감사합니다^^

injae7034 committed:
코드 리뷰 개선

injae7034 commented:
# 코드리뷰 받은 후 개선사항
## 1. HelloController에 대한 JavaDoc 주석 달기
## 2. App클래스에서 getGreeting 메소드 제거 후 AppTest 클래스 제거
이후 테스트 코드를 실행시켰지만 아직 까지 별다른 문제를 찾지 못했음
## 3. Context 스코프의 역할
Context에서 Test에 사용할 id 변수 또는 title변수들을 미리 final로 선언해줌.
## 4. given을 사용하지 않고 TaskControllerMock_MVC_Test를 테스트할 방법 고찰
아직 어떻게 해야할지 모르겠음.<br>
자칫 잘못하면 TaskControllerTest와 중복이 될 거 같은데<br>
어떻게 해야 TaskControllerMock_MVC_Test와 TaskControllerTest를 구별해서 테스트할 수 있을지 아직 실마리를 찾지 못함.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/HelloController.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * HelloController 클래스는 클라이언트와 서버가 제대로 연결되었는지 확인하기 위한 테스트 컨트롤러의 역할을 합니다.<br>
+ *
+ * 간단하고 빠르게 테스트하기 위해서 Hello, world!가 화면에 잘출력되는지 확인합니다.<br>
+ *
+ * 잘출력된다면 서버와 클라이언트가 서로 제대로 연결이 되었다는 것을 의미합니다.<br>
+ *
+ * 제대로 출력이 되지 않는다면 서버 또는 연결에 문제가 있다는 것을 의미합니다.<br>
```
내용은 완벽합니다. 그런데 좀 길군요! 이 네 줄을 한 줄로 축약해 보세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,142 @@
+            }
+        }
+
+        @Nested
+        @DisplayName("클라이언트가 요청한 Task의 id가 Tasks에 없으면")
+        class Context_with_invalid_id {
+            final long INVALID_REQUEST_TASK_ID = 100L;
+
+            @Test
+            @DisplayName("TaskNotFoundException 예외를 던집니다.")
```
컨텍스트의 설명에 대해 없는 아이디로 `100L`을 선언했는데요, 논리적인 근거가 없습니다. `100L`이 tasks에 없는 상황을 코드를 통해서 이 컨텍스트 안에 표현해 보세요. 이건 어떻게 보면 아주 쉽고 어떻게 보면 아주 어려울 수 있습니다. 열심히 생각해서 작업해 보세요.

한편 컨텍스트가 테스트를 준비하는 공간이라는 것을 꼭 기억해 두세요. 테스트를 준비한다는 것은 숫자를 미리 마련해두는 정도가 아닐 수 있습니다. "Task의 id가 Tasks에 없도록" 상황을 만들어주는 것이 중요합니다.

injae7034 committed:
TaskControllerMock_MVC_Test 성공

injae7034 commented:
# 코드 리뷰 개선
## 1. HelloController의 JavaDoc 주석 1줄로 축약함.
## 2. given은 사용하지 않고, MockMvc를 사용하여 TaskControllerMock_MVC_Test하기
MockMvc를 사용하면 반드시 given을 사용해야 하는 줄 알았는데 그게 아니었습니다.<br>
일단 TaskService에서 MockBean애노테이션을 제거하고 MockMvc에서도 Autowired어노테이션을 제거한 다음<br>
setUp에서 직접 할당을 해줬습니다.<br>
이 후 CRUD의 각 Context에서 필요한 상황에 맞게 TaskService에 task를 추가하고,<br>
MockMvc를 컨트롤러로 활용하여 perform을 이용해 테스트하였고,<br>
CRUD의 모든 상황에 대해서 테스트를 통과하였습니다.<br>
다만 모든 Tasks를 읽는 항목 테스트에서 10개의 task를 TaskService에 넣었는데 이 10개가 다 잘들어갔는지 테스트하기 위해<br>10개 항목을 모두 반복문없이 검사하고 있는데 이를 반복문으로 할 방법을 아직 찾지 못했습니다.<br>
스트림을 사용하여 forEach를 하면 뭔가 한줄에 해결이 될거같기도 한데 아직은 해결하지 못해서 10줄 노가다를 하였습니다;;<br>
## 3.  Context에서 논리적으로 상황만들기
"클라이언트가 요청한 Task의 id가 Tasks에 없으면" 에서 tasks에 없는 상황을 코드를 통해서 컨텍스트 안에 표현해보기<br>
1. Context의 setUp에서 비어있는 taskService에 10개의 task를 채워넣음.<br>
2. 그러므로 현재 tasksSize는 10개임.
3. 11번째 id를 조회함.(detail)
4. size가 10개인데 size보다 큰 id를 조회하기 때문에 TaskNotFoundException이 발생함.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMock_MVC_Test.java:
```diff
@@ -82,14 +76,49 @@ void create() throws Exception {
+                @Test
+                @DisplayName("할 일 목록에 있는 모든 Task들을 반환합니다.")
+                void list() throws Exception {
+                    mockMvc.perform(get("/tasks"))
+                            .andExpect(status().isOk())
+                            .andExpect(content().string(containsString(TASK_TITLE + 1)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 2)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 3)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 4)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 5)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 6)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 7)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 8)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 9)))
+                            .andExpect(content().string(containsString(TASK_TITLE + 10)));
+                }
```
이런 경우에는 적당히 타협하는 것도 방법 중 하나입니다. 😅
또는 json으로 변환해서 미리 변환해 둔 json 하고 비교해보는 방법도 있겠죠.

injae7034 replied:
Task에 taskToJson이라는 메소드를 생성하고, TaskService에 tasksToJson을 생성하여 내부에서 Task의 taskToJson을 호출하여 json형식으로 10개의 task를 변경한 다음 비교하는 방법을 사용하여 테스트를 통과하였습니다.
```java
 @Test
@DisplayName("할 일 목록에 있는 모든 Task들을 반환합니다.")
void list() throws Exception {
         mockMvc.perform(get("/tasks"))
                  .andExpect(status().isOk())
                  .andExpect(content().string(equalTo(taskService.tasksToJson())));
}
```
이 방법은 괜찮을까요?

injae7034 replied:
코드를 pull request하고 나서 생각해보니 1주차때 Task와 Tasks에 toJson으로 변환하는 메소드를 분리시켜라고 하셨던 것이 생각나서 PrinterService라는 클래스를 새로 만들어서 여기서 taskToJson과 tasksToJson메소드를 static으로 정의하였습니다.<br>
원래는 Json으로만 변경하는 메소드밖에서 없어서 클래스명을 JsonPrinterService라고 정의하려고 하였으나 나중에 xml이나 다른 형식의 출력이 필요하면 추가로 여기에 해당 양식을 출력할 메소드를 정의하는 것이 좋을 것 같아서 클래스 이름을 JsonPrinterService보다는 PrinterService로 명명하였습니다.<br>
혹시 이것이 너무 앞서서 생각하는 오버엔지니어링에 속하는 것일까요?<br>
(YAGNI원칙에 위배되는 것 같다는 생각이 갑자기 들어서요;;)

johngrib replied:
오버엔지니어링이면 어떻습니까? 인재님은 이번에 각 코드의 역할과 책임을 의식해서 분리하는 연습을 하신 겁니다. 이렇게 의식적으로 역할을 나누고 코드를 쪼개는 것은 아주 훌륭한 객체지향 공부이기도 합니다. 더 열심히 고민해 보세요.

johngrib reviewed (commented):
좀 어려운 중간 과제를 드렸는데 열심히 노력해서 잘 해내신 것 같습니다!
예상보다 잘 하셔서 저까지 기분이 좋습니다.

물론 아직 더 고쳐야 할 점들도 꽤 있긴 한데 점점 배워나가는 과정이니 앞으로 하나하나 흡수해가시면 될 거 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/HelloController.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * HelloController 클래스는 클라이언트와 서버가 제대로 연결되었는지 확인하기 위한 테스트 컨트롤러의 역할을 합니다.<br>
+ */
+
+@RestController
```
이 정도로 표현해도 괜찮겠죠.
```suggestion
/**
 * 클라이언트와 서버가 제대로 연결되었는지 확인하는 역할을 합니다.
 */
@RestController
```

injae7034 replied:
군더더기없이 훨씬 간단하고 명료해졌네요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,142 @@
+        @DisplayName("Tasks에 있는 모든 Task를 반환합니다.")
+        void getTasks() {
+            assertThat(taskService.getTasks()).isNotEmpty();
+            assertThat(taskService.getTasks()).hasSize(1);
```
테스트를 잘 작성하셨지만 아직 더 고쳐야 할 점이 있습니다. `.hasSize(1)`을 봅시다. 이 `1`은 확정된 상수여서 변경에 취약해 보입니다. 30번 라인을 보면 task를 하나 추가하고 있죠? 그런데 누가 `@BeforeEach`를 `task` 3개를 추가하도록 고쳐놓고 확인도 안 하고 어디 갔다고 생각해 봅시다. 그러면 이 테스트는 실패할 것입니다.

injae7034 replied:
TASKS_SIZE = 5로 상수를 선언하고 setUP에서 5개만큼 반복을 돌려 task를 채워 넣은 다음에 테스트메소드에서 hasSize(TASKS_SIZE )로 한다면 괜찮을까요?

johngrib replied:
좋은 방법들 중의 하나입니다. 그렇게 하는 것도 괜찮죠.

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerMock_MVC_Test.java:
```diff
@@ -0,0 +1,230 @@
+        @Nested
+        @DisplayName("list 메소드는")
+        class Describe_list {
+            @Nested
+            @DisplayName("처음에는")
+            class Context_at_first {
```
`처음에는`이라는 표현이 애매합니다. 더 좋은 표현을 생각해 봅시다.

injae7034 replied:
"Tasks에 할 일 목록이 하나도 없으면"으로 수정하였습니다.

injae7034 committed:
코드 리뷰 개선

injae7034 committed:
PrinterService 클래스 생성

injae7034 commented:
## 코드리뷰 개선 사항
### 1. TaskControllerMock_MVC_Test에서 list에 테스트 개선
TaskServie에 10개의 항목을 채운다음 이를 list 메소드테스트에서 확인하기 위해<br>
1~10까지 다 있는지 확인하기 위해 10줄을 작성했는데 굉장히 비효율적임.<br>
그래서 처음에는 Task에서 taskToJson메소드를 생성하고,<br>
TaskService에서 tasksToJson메소드를 새로 정의함.<br>
하지만 1주차 때 리뷰해주신 내용 중에 task에서 json 양식으로 변경하는 역할을 분리하라고 했던 것이 생각남.<br>
그래서 PrinterService 클래스를 새로 만들고 여기에서 taskToJson과 tasksToJson 메소드를 정의함.<br>
JsonPrinterService보다 PrinterService로 명명한 이유는 지금은 json양식 하나뿐이지만<br>
나중에 task를 다른 양식으로 출력할 일이 생겼을 때<br>
그 때 마다 PrinterService 를 따로 생성하는 것보다는 출력과 관련된 처리는 모두 PrinterService 에서 해주기 위해<br>
범용적인 이름을 지음.<br>
그러나 이것이 자칫 잘못하면 오버엔지니어링이 될 수도 있을 것 같다는 생각이 들어 그 경계를 아직 잘모르겠음.

### 2. HelloController JavaDoc 한줄 내용 변경

### 3. TaskServiceTest에서 getTasks 메소드 테스트 취약점 개선
TASKS_SIZE = 5로 상수를 선언하고 setUP에서 5개만큼 반복을 돌려 task를 채워 넣은 다음에 테스트 메소드에서 hasSize(TASKS_SIZE )를 이용하여 테스트하도록 변경

### 4. TaskControllerMock_MVC_Test list의 context 상황 설명 변경
"처음에는"이라는 표현이 애매하여 "Tasks에 할 일 목록이 하나도 없으면"으로 수정하였습니다.

injae7034 commented:
# \@Autowired 질문사항
스프링에 대해 아직 모르는 것이 너무 많아서 최근에 혼자서 조금 공부를 하고 있습니다.<br>
최근에 \@Component와 \@ComponentScan, \@Autowired에 대해서 알게 되었습니다.<br>
현재 TaskController 클래스는 생성자에서 taskService를 주입받고 있습니다.
## 기존 TaskController 코드
```java
@RestController
@RequestMapping("/tasks")
@CrossOrigin
public class TaskController {
    private TaskService taskService;

    public TaskController(TaskService taskService) {
        this.taskService = taskService;
    }
}
```
TaskService는 \@Service로 인해 스프링빈의 검색 대상에 등록이 됩니다.
## TaskService 코드
```java
@Service
public class TaskService {
    private List<Task> tasks = new ArrayList<>();
    private Long newId = 0L;
}
```

그래서 TaskController에서 \@Autowired를 통해 스프링이 자동으로 TaskService에 대한 의존주입을 해주도록 바꿨습니다.<br>
그럼 더이상 생성자에서 의존주입을 할 필요가 없으니 (기본)생성자 정의를 없앴습니다. 
## 변경된 TaskController 코드
```java
@RestController
@RequestMapping("/tasks")
@CrossOrigin
public class TaskController {
    @Autowired
    private TaskService taskService;
}
```
앱에서는 \@SpringBootApplication이 있고 여기에 \@ComponentScan이 있기 때문에<br>
App은 정상적으로 돌아갈 것이라 예상되고 실제로 콘솔창에 실험을 해보니 정상적으로 돌아갑니다.<br>
## App 코드
```java
@SpringBootApplication
public class App {
    public static void main(String[] args) {
        SpringApplication.run(App.class, args);
    }
}
```
## 테스트 코드
TaskControllerTest의 경우 기존코드는 아래와 같습니다.
setUp에서 TaskController에 의존 주입을 해주고 있습니다.
## 기존 TaskControllerTest 코드
```java
@DisplayName("TaskController 클래스")
class TaskControllerTest {
    private static final String TASK_TITLE_ONE = "testOne";
    private static final String TASK_TITLE_TWO = "testTwo";
    private static final String UPDATE_TITLE = "otherTest";

    private TaskController controller;
    private TaskService taskService;

    @BeforeEach
    void setUp() {
        taskService = new TaskService();
        controller = new TaskController(taskService);

        Task task = new Task();
        task.setTitle(TASK_TITLE_ONE);
        controller.create(task);
    }
}
```
이를 아래처럼 바꿨습니다. Autowired를 통해 스프링이 자동으로 의존주입을 해주기 때문에 더이상
생성자를 호출하지 않았습니다.<br>
ComponentScan이 없어서 안되나 싶어 SpringBootApplication 애노테이션을 임시적으로 사용하였습니다.
## 변경된 TaskControllerTest 코드
```java
@SpringBootApplication
@DisplayName("TaskController 클래스")
class TaskControllerTest {
    private static final String TASK_TITLE_ONE = "testOne";
    private static final String TASK_TITLE_TWO = "testTwo";
    private static final String UPDATE_TITLE = "otherTest";

    @Autowired
    private TaskController controller;
    @Autowired
    private TaskService taskService;

    @BeforeEach
    void setUp() {
        Task task = new Task();
        task.setTitle(TASK_TITLE_ONE);
        controller.create(task);
    }
}
```

TaskControllerMock_MVC_Test 의 경우 처음에는 아래와 같이 작성하여<br>
mockMvc이 생성될 때 TaskController에 TaskService를 의존주입해주고 있습니다.
## 기존 TaskControllerMock_MVC_Test  코드
```java
@SpringBootTest
@AutoConfigureMockMvc
public class TaskControllerMock_MVC_Test {
    private static final String TASK_TITLE_ONE = "testOne";

    MockMvc mockMvc;

    private TaskService taskService;

    @BeforeEach
    void setUp() {
        taskService = new TaskService();

        this.mockMvc = MockMvcBuilders
                .standaloneSetup(new TaskController(taskService))
                .setControllerAdvice(TaskErrorAdvice.class)
                .build();
    }
```
이후 TaskController의 taskService에 Autowired 애노테이션을 선언해주고 생성자에서 의존주입하는 것을 없앴습니다.<br>
## 변경된 TaskControllerMock_MVC_Test 코드
```java
@SpringBootTest
@AutoConfigureMockMvc
public class TaskControllerMock_MVC_Test {
    private static final String TASK_TITLE_ONE = "testOne";

    MockMvc mockMvc;

    private TaskService taskService;

    @BeforeEach
    void setUp() {
        taskService = new TaskService();

        this.mockMvc = MockMvcBuilders
                .standaloneSetup(new TaskController())
                .setControllerAdvice(TaskErrorAdvice.class)
                .build();
    }
}
```
## 변경된 테스트 코드 결과
두 테스트의 결과는 모든 테스트가 NullPointerException이 발생했습니다.<br>
즉, 아예 생성이 안되었다는 말인데 이유를 잘모르겠습니다.<br><br>
혹시 App에서는 AutoWired가 잘되어 정상적으로 잘돌아가는데<br>
테스트에서는 왜 안돌아가는지 이유를 알 수 있을까요?<br>
SpringBootTest 애노테이션에 ComponentScan이 없어서 안되나 싶어<br>
SpringBootAplication을 임시방편으로 사용했으나 테스트에 실패하였습니다ㅠ

johngrib commented:
![image](https://user-images.githubusercontent.com/1855714/162576346-3ab1e8de-0dc6-4044-8f29-c88f0c85cf20.png)

⬆ 이렇게 하지 않는 것이 좋습니다. 가급적이면 `@Autowired`를 사용하지 마세요. 생성자를 사용하는 방법으로 다시 고치시길 바랍니다.
(테스트 파일 정도는 `@Autowired`를 사용해도 괜찮지만 가급적이면 도메인 코드에서는 `@Autowired`를 사용하지 마세요.)
`@Autowired`를 사용하지 않고 생성자를 사용하는 이유에 대해 조사하고 공부해 보신 다음 글로 작성해서 정리해 보세요.

그리고 테스트가 실패했다고 하셨는데요, 어떤 에러 메시지가 나왔는지 읽고 파악하는 것이 순서입니다.
제가 받아서 돌려보니 다 문제 없이 잘 돌아가더라고요.
NullPointerException이 발생하는 이유는 다양하고, 저는 이 설명만으로는 어떤 상황인지 단언하기 어렵습니다.
에러 메시지를 읽고 상황에 대해 고민해 보세요. 질문을 할 때에도 에러 메시지를 같이 올리는 것이 기본입니다.

johngrib commented:
어느새 일요일이 되었네요. 한 주간 어떠셨나요? 주간 회고를 작성해 공유해 주세요.

injae7034 committed:
Update TaskControllerMock_MVC_Test.java
